ESMF/UFS/Aerosol_Comp_Mod.F90

```diff
@@ -709,6 +720,101 @@ subroutine AerosolTracerUpdate(state, cap, rc)
 
   end subroutine AerosolTracerUpdate
 
+  subroutine AerosolDiagUpdate(model, cap, rc)
+    type(ESMF_GridComp)            :: model
+    type(MAPL_Cap)                 :: cap
+    integer, optional, intent(out) :: rc
+
+    ! -- local variables
+    integer :: localrc
+    integer :: i, j, k, kk, ni, nj, nk
+    real(ESMF_KIND_R4) :: pm25, pm10
+    real(ESMF_KIND_R4), dimension(:,:,:),   pointer :: rho
+    real(ESMF_KIND_R8), dimension(:,:,:,:), pointer :: q
+
+    type(ESMF_Field) :: field
+    type(ESMF_State) :: state
+
+    ! -- local parameters
+    real(ESMF_KIND_R8), parameter :: w_du2  = log(1.250_ESMF_KIND_R8) / log(1.8_ESMF_KIND_R8)
+    real(ESMF_KIND_R8), parameter :: w_du4  = log(1.667_ESMF_KIND_R8) / log(2.0_ESMF_KIND_R8)
+    real(ESMF_KIND_R8), parameter :: w_ss3  = log(2.50_ESMF_KIND_R8) / log(3._ESMF_KIND_R8)
+    real(ESMF_KIND_R8), parameter :: w_so4  = 132.14_ESMF_KIND_R8 / 96.06_ESMF_KIND_R8
+    real(ESMF_KIND_R8), parameter :: w_no3  = 80.043_ESMF_KIND_R8 / 62.0_ESMF_KIND_R8
+    real(ESMF_KIND_R8), parameter :: w25_no3an2 = 0.138_ESMF_KIND_R8 * w_no3
+    real(ESMF_KIND_R8), parameter :: w10_no3an2 = 0.808_ESMF_KIND_R8 * w_no3
+    real(ESMF_KIND_R8), parameter :: w10_no3an3 = 0.164_ESMF_KIND_R8 * w_no3
+
+    ! -- begin
+    if (present(rc)) rc = ESMF_SUCCESS
+
+    ! -- retrieve export state
+    call ESMF_GridCompGet(model, exportState=state, rc=localrc)
+    if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, &
+      line=__LINE__,  &
+      file=__FILE__,  &
+      rcToReturn=rc)) return  ! bail out
+
+    ! -- retrieve tracers
+    nullify(q)
+    call AerosolGetPtr(state, "inst_tracer_mass_frac", q, rc=localrc)
+    if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, &
+      line=__LINE__,  &
+      file=__FILE__,  &
+      rcToReturn=rc)) return  ! bail out
+
+    ! -- retrieve air density
+    call ESMF_StateGet(cap % cap_gc % import_state, "AIRDENS", field, rc=localrc)
+    if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, &
+      line=__LINE__,  &
+      file=__FILE__,  &
+      rcToReturn=rc)) return  ! bail out
+
+    nullify(rho)
+    call ESMF_FieldGet(field, farrayPtr=rho, rc=localrc)
+    if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, &
+      line=__LINE__,  &
+      file=__FILE__,  &
+      rcToReturn=rc)) return  ! bail out
+
+    ! -- compute pm
+    ni = size(q,1)
+    nj = size(q,2)
+    nk = size(q,3)
+
+    do k = 1, nk
+      kk = nk - k + 1
+      do j = 1, nj
+        do i = 1, ni
+          ! -- compute partial PM2.5
+          pm25 = q(i,j,k,p_bc_1) + q(i,j,k,p_bc_2)     &
+               + q(i,j,k,p_oc_1) + q(i,j,k,p_oc_2)     &
+               + q(i,j,k,p_dust_1)                     &
+               + q(i,j,k,p_seas_1) + q(i,j,k,p_seas_2) &
+               + w_so4 * q(i,j,k,p_sulf)               &
+               + w_no3 * q(i,j,k,p_no3an1)
```

Comment:
1. @gmao-esherman Would this work if one or more aerosol types are disabled, dust for example?
2. @gmao-esherman I would expect that the PM diagnostics is calculated using only the active aerosols, is this the case?

Reply:
This is all under ESMF/UFS (NOAA stuff). Nothing under ESMF/UFS gets exercised by legacy GOCART or GOCART2G when we build GEOS.

Reply:
@gmao-esherman  Why are we asked to review it then, just a formality?

Reply:
Good question. I am not 100% sure why. I think any changes made to the @GOCART repository have to get reviewed by someone. Maybe @tclune has some more insight.

Reply:
@adarmenov, @gmao-esherman - Good point. As long as all required aerosol tracers are included in the atmospheric model, the PM computation will not be affected by one or more aerosol types being disabled. More flexibility might be added in the future.